### PR TITLE
gitea#3868 -- Implement /api/indexer/info in js-client

### PR DESCRIPTION
**Pull request by Puszkarek**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,6 +43,7 @@
 		"@types/node": "14.14.27",
 		"base-64": "^1.0.0",
 		"date-fns": "^2.17.0",
+		"decoders": "~1.22.2",
```
Comment (iagojsilva) [inline package.json:46]: S: If this version is the only compatible with the project, remove the `~`

File: src/functions/indexers/get-all-indexers.spec.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ **************************************************************************/
+
+import { isIndexerWell } from '~/models';
+import { integrationTest, TEST_BASE_API_CONTEXT } from '~/tests';
+import { makeGetAllIndexers } from '.';
+
+describe('getAllIndexers()', () => {
+	const getAllIndexers = makeGetAllIndexers(TEST_BASE_API_CONTEXT);
+
+	it(
```
Comment (iagojsilva) [inline src/functions/indexers/get-all-indexers.spec.ts:16]: S: Create another test: `Should return an empty array if there are no Indexers`

File: src/models/indexer/is-indexer-well.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+import { isArray, isObject, isString } from 'lodash';
+import { isUUID } from '~/value-objects';
+import { IndexerWell } from './indexer-well';
+
+export const isIndexerWell = (value: any): value is IndexerWell => {
+	try {
+		const i = <IndexerWell>value;
+
+		return isUUID(i.UUID) && isString(i.Name) && isArray(i.Wells) && isObject(i.Replicated);
```
Comment (iagojsilva) [inline src/models/indexer/is-indexer-well.ts:17]:
```
const value = {
 UUID: 1,
 Name: 'well-name'
 Wells: ['i don't care what I am']
 Replicated: ['i'm an object too']
}
isIndexerWell(value) // returns true

```
M: Improve this check. Take a look at `is-saved-query.ts`

File: src/models/indexer/response-guard.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { guard } from 'decoders';
+import { indexerWellDecoder } from './indexer-well-decoder';
+
+export const responseGuard = guard(indexerWellDecoder);
```
Comment (iagojsilva) [inline src/models/indexer/response-guard.ts:12]: S: Move this to the same file of `indexerWellDecoder` to improve reading.

File: src/models/indexer/is-indexer-well.ts (hunk 3 above)
Comment (Puszkarek) [reply src/models/indexer/is-indexer-well.ts:17]: tidy

File: src/functions/indexers/get-all-indexers.spec.ts (hunk 2 above)
Comment (Puszkarek) [reply src/functions/indexers/get-all-indexers.spec.ts:16]: for that, it would not be necessary to have the `deleteOneIndexer` method? like in `get-all-groups.spec.ts`

File: src/models/indexer/response-guard.ts (hunk 4 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.ts:12]: tidy

File: package.json (hunk 1 above)
Comment (iagojsilva) [reply package.json:46]: We don't know if they release a new minor version it'll break or not, I got you probably not, but remove the `~` for safety if we need we upgrade it manually

File: src/functions/indexers/get-all-indexers.spec.ts (hunk 2 above)
Comment (iagojsilva) [reply src/functions/indexers/get-all-indexers.spec.ts:16]: ok

File: src/models/indexer/indexer-well.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+/** An indexer's well. */
+export type IndexerWell = {
+	Name: string;
+	UUID: string;
+	Wells: Array<Well>;
+	Replicated?: Record<string, Array<ReplicatedState>>;
+};
+
+export const isIndexerWell = (value: any): value is IndexerWell => {
```
Comment (iagojsilva) [inline src/models/indexer/indexer-well.ts:22]: S: Use `unknown` instead of `any`

File: src/models/indexer/replicated-state.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface ReplicatedState {
+	Name: string;
+	Accelerator?: string;
+	Engine?: string;
+	Tags: Array<string>;
+	Shards: Array<Shard>;
+}
+
+export const isReplicatedState = (value: any): value is ReplicatedState => {
```
Comment (iagojsilva) [inline src/models/indexer/replicated-state.ts:20]: S: Use `unknown` instead of `any`

File: src/models/indexer/shard.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	Size: number;
+	Cold: boolean;
+	RemoteState?: {
+		UUID: string;
+		Entries: number;
+		Size: number;
+	};
+};
+
+export const isShard = (value: any): value is Shard => {
```
Comment (iagojsilva) [inline src/models/indexer/shard.ts:25]: S: Use `unknown` instead of `any`

File: src/models/indexer/well.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export type Well = {
+	Name: string;
+	Accelerator?: string;
+	Engine?: string;
+	Path: string;
+	Tags: Array<string>;
+	Shards: Array<Shard>;
+};
+
+export const isWell = (value: any): value is Well => {
```
Comment (iagojsilva) [inline src/models/indexer/well.ts:21]: S: Use `unknown` instead of `any`

File: src/models/indexer/shard.ts (hunk 7 above)
Comment (Puszkarek) [reply src/models/indexer/shard.ts]: done

File: src/models/indexer/replicated-state.ts (hunk 6 above)
Comment (Puszkarek) [reply src/models/indexer/replicated-state.ts]: done

File: src/models/indexer/indexer-well.ts (hunk 5 above)
Comment (Puszkarek) [reply src/models/indexer/indexer-well.ts]: done

File: src/models/indexer/response-guard.ts (hunk 4 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.ts:54]: done

File: src/models/indexer/well.ts (hunk 8 above)
Comment (Puszkarek) [reply src/models/indexer/well.ts]: done

File: src/functions/indexers/get-all-indexers.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+export const makeGetAllIndexers = (context: APIContext) => {
+	const templatePath = '/api/indexer/info';
+	const url = buildURL(templatePath, { ...context, protocol: 'http' });
+
+	return async (): Promise<Array<IndexerWell>> => {
+		const req = buildHTTPRequestWithAuthFromContext(context);
+
+		const raw = await fetch(url, { ...req, method: 'GET' });
+		const rawRes = (await parseJSONResponse<any>(raw)) ?? [];
```
Comment (LucasPaganini) [inline src/functions/indexers/get-all-indexers.ts:20]: M: Don't use `<any>`, use the correct expected type

File: src/functions/indexers/get-all-indexers.spec.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * Copyright 2021 Gravwell, Inc. All rights reserved.
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { isIndexerWell } from '~/models';
+import { integrationTest, TEST_BASE_API_CONTEXT } from '~/tests';
+import { makeGetAllIndexers } from '.';
```
Comment (LucasPaganini) [inline src/functions/indexers/get-all-indexers.spec.ts:11]: M: Don't import from `'.'`, import from the actual file

File: src/functions/indexers/get-all-indexers.spec.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+describe('getAllIndexers()', () => {
+	const getAllIndexers = makeGetAllIndexers(TEST_BASE_API_CONTEXT);
+
+	it(
+		'Should return all IndexersWells',
```
Comment (LucasPaganini) [inline src/functions/indexers/get-all-indexers.spec.ts:17]: S: Typo, `IndexersWells` should be `IndexerWells`. The `s` comes at the end.

File: src/models/indexer/indexer-well.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { isString, isUndefined } from 'lodash';
+import { isUUID } from '~/value-objects';
+import { isReplicatedState, ReplicatedState } from './replicated-state';
+import { isWell, Well } from './well';
+
+/** An indexer's well. */
+export type IndexerWell = {
```
Comment (LucasPaganini) [inline src/models/indexer/indexer-well.ts:15]: M: This is the `RawIndexerWell`. Create another interface, normalized, with better parameter names, to be the `IndexerWell`.

File: src/models/indexer/replicated-state.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { isString, isUndefined } from 'lodash';
+import { isShard, Shard } from './shard';
+
+export interface ReplicatedState {
```
Comment (LucasPaganini) [inline src/models/indexer/replicated-state.ts:12]: M: This is the `RawReplicatedState`. Create another interface, normalized, with better parameter names, to be the `ReplicatedState`.

File: src/models/indexer/response-guard.spec.ts [hunk 14]
```diff
@@ -0,0 +1,200 @@
+/*************************************************************************
+ * Copyright 2021 Gravwell, Inc. All rights reserved.
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { responseGuard } from '.';
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.spec.ts:9]: M: Don't import from `'.'`

File: src/models/indexer/response-guard.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+ * Copyright 2021 Gravwell, Inc. All rights reserved.
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { array, boolean, dict, guard, iso8601, number, object, optional, string } from 'decoders';
+
+const shardDecoder = object({
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.ts:11]: S: Rename to `rawShardDecoder`

File: src/models/indexer/response-guard.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	RemoteState: optional(
+		object({
+			UUID: string,
+			Entries: number,
+			Size: number,
+		}),
+	),
+});
+
+const indexerWellDecoder = dict(
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.ts:27]: S: Rename to `rawIndexerWellDecoder`

File: src/models/indexer/response-guard.ts [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+					Engine: optional(string),
+					Tags: array(string),
+					Shards: array(shardDecoder),
+				}),
+			),
+		),
+	}),
+);
+
+export const responseGuard = guard(indexerWellDecoder);
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.ts:54]: M: Rename to `assertIsRawIndexerWell`

File: src/models/indexer/response-guard.spec.ts [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+fdescribe('responseGuard()', () => {
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.spec.ts:11]: M: Don't `fdescribe`

File: src/models/indexer/response-guard.spec.ts [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+	describe('Should work', () => {
```
Comment (LucasPaganini) [inline src/models/indexer/response-guard.spec.ts:12]: M: Add a test for the function types. It should have the type `(data: unknown) => asserts data is RawIndexerWell`

File: src/models/indexer/shard.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+ * Copyright 2021 Gravwell, Inc. All rights reserved.
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { has, isBoolean, isDate, isNumber, isString, isUndefined } from 'lodash';
+
+export type Shard = {
```
Comment (LucasPaganini) [inline src/models/indexer/shard.ts:11]: M: This is the `RawShard`. Create another interface, normalized, with better parameter names, to be the `Shard`.

File: src/models/indexer/well.ts [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+ * Contact: <legal@gravwell.io>
+ *
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { isString, isUndefined } from 'lodash';
+import { isShard, Shard } from './shard';
+
+export type Well = {
```
Comment (LucasPaganini) [inline src/models/indexer/well.ts:12]: M: This is the `RawWell`. Create another interface, normalized, with better parameter names, to be the `Well`.

**Review by LucasPaganini — CHANGES_REQUESTED** [review]

File: src/functions/indexers/get-all-indexers.ts (hunk 9 above)
Comment (Puszkarek) [reply src/functions/indexers/get-all-indexers.ts:20]: tidy

File: src/functions/indexers/get-all-indexers.spec.ts (hunk 10 above)
Comment (Puszkarek) [reply src/functions/indexers/get-all-indexers.spec.ts:11]: tidy

File: src/functions/indexers/get-all-indexers.spec.ts (hunk 11 above)
Comment (Puszkarek) [reply src/functions/indexers/get-all-indexers.spec.ts:17]: tidy

File: src/models/indexer/indexer-well.ts (hunk 12 above)
Comment (Puszkarek) [reply src/models/indexer/indexer-well.ts:15]: I think that this is the IndexerWell and the RawIndexerWell is some like
```
export type RawIndexerWell = {
	[key: string]: Omit<IndexerWell, 'Name'>;
};
```

File: src/models/indexer/well.ts (hunk 21 above)
Comment (Puszkarek) [reply src/models/indexer/well.ts:12]: but this decoder also converts `iso8601` string from
const shardDecoder = object({
	Start: iso8601,
	End: iso8601,
	...
	),
to `Date`, I don't think that `asserts` would be the best option

File: src/models/indexer/response-guard.ts (hunk 17 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.ts:54]: but this decoder also converts `iso8601` string from
const shardDecoder = object({
	Start: iso8601,
	End: iso8601,
	...
	),
to `Date`, I don't think that `asserts` would be the best option

File: src/models/indexer/response-guard.ts (hunk 15 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.ts:11]: tidy

File: src/models/indexer/response-guard.ts (hunk 16 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.ts:27]: tidy

File: src/models/indexer/response-guard.spec.ts (hunk 14 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts:9]: tidy

File: src/models/indexer/replicated-state.ts (hunk 13 above)
Comment (Puszkarek) [reply src/models/indexer/replicated-state.ts:12]: tidy

File: src/models/indexer/response-guard.spec.ts (hunk 18 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts:11]: tidy

File: src/models/indexer/shard.ts (hunk 20 above)
Comment (Puszkarek) [reply src/models/indexer/shard.ts:11]: tidy

File: src/models/indexer/well.ts (hunk 21 above)
Comment (Puszkarek) [reply src/models/indexer/well.ts:12]: tidy

File: src/models/indexer/response-guard.spec.ts (hunk 19 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts:12]: `asserts` is not compatible with typescript v3.5 and 3.6

File: src/models/indexer/shard.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+	End: Date;
+	Entries: number;
+	Size: number;
+	Cold: boolean;
+	RemoteState?: {
+		UUID: string;
+		Entries: number;
+		Size: number;
+	};
+};
```
Comment (iagojsilva) [inline src/models/indexer/shard.ts:23]: S: Use camelCase in the normalized types

File: src/models/indexer/well.ts [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+	Name: string;
+	Accelerator?: string;
+	Engine?: string;
+	Path: string;
+	Tags: Array<string>;
+	Shards: Array<Shard>;
+};
```
Comment (iagojsilva) [inline src/models/indexer/well.ts:19]: S: Use camelCase in the normalized types

File: src/models/indexer/replicated-state.ts (hunk 13 above)
Comment (iagojsilva) [inline src/models/indexer/replicated-state.ts:12]: S: Use camelCase in the normalized types

File: src/models/indexer/indexer-well.ts [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+	name: string;
+	UUID: string;
```
Comment (iagojsilva) [inline src/models/indexer/indexer-well.ts:17]: M: use `uuid` instead of `UUID`

File: src/models/indexer/raw-indexer-well-decoder.ts [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { dict, guard, object, string } from 'decoders';
+import { RawReplicatedStateDecoded, rawReplicatedStateDecoder } from './raw-replicated-state-decoder';
+import { RawWellDecoded, rawWellDecoder } from './raw-well-decoder';
+
+export type RawIndexerWellDecoded = {
+	UUID: string;
```
Comment (iagojsilva) [inline src/models/indexer/raw-indexer-well-decoder.ts:14]: M: use uuid instead of UUID

File: src/models/indexer/raw-indexer-well.ts [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+ * This software may be modified and distributed under the terms of the
+ * MIT license. See the LICENSE file for details.
+ **************************************************************************/
+
+import { RawReplicatedState } from './raw-replicated-state';
+import { RawWell } from './raw-well';
+
+export type RawIndexerWell = {
+	[key: string]: {
+		UUID: string;
```
Comment (iagojsilva) [inline src/models/indexer/raw-indexer-well.ts:14]: M: use uuid instead of UUID

File: src/models/indexer/raw-replicated-state.ts [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+ **************************************************************************/
+
+import { RawShardDecoded } from './raw-shard-decoder';
+
+export interface RawReplicatedState {
+	Name: string;
+	Accelerator?: string;
+	Engine?: string;
+	Tags: Array<string>;
+	Shards: Array<RawShardDecoded>;
```
Comment (iagojsilva) [inline src/models/indexer/raw-replicated-state.ts:16]: S: Why this is `RawShardDecoded` and not `RawShard`

File: src/models/indexer/response-guard.spec.ts [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+		});
+		it('Should FAILED errors if properties are INVALIDs', () => {
+			const { withInvalidProperty } = invalidMockData();
+
+			expect(() => rawIndexerWellGuard(withInvalidProperty)).toThrowError();
+		});
+	});
+});
+
+const validMockData = () => {
```
Comment (iagojsilva) [inline src/models/indexer/response-guard.spec.ts:48]: S: Improve your mocks, make them strictly typed and there are too big there is no need to set all these props just set the required ones

File: src/models/indexer/response-guard.spec.ts [hunk 29]
```diff
@@ -0,0 +1,200 @@
...
+	const validEmpty = {};
```
Comment (iagojsilva) [inline src/models/indexer/response-guard.spec.ts:49]: S: I still don't get how an empty object could be valid data for a type whose props are all mandatory

Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts:49]: because the object is inside a key `{[key: string]: Indexer}`

ex:
```
const valid = {}
const invalid = {
	'name-key': undefined
}
```
like an empty array

File: src/models/indexer/indexer-well.ts (hunk 24 above)
Comment (Puszkarek) [reply src/models/indexer/indexer-well.ts:17]: tidy

File: src/models/indexer/raw-replicated-state.ts (hunk 27 above)
Comment (Puszkarek) [reply src/models/indexer/raw-replicated-state.ts]: fixed

File: src/models/indexer/response-guard.spec.ts (hunk 28 above)
Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts]: I made some improvements to the mock, but most properties are required

File: src/models/indexer/response-guard.spec.ts (hunk 29 above)
Comment (iagojsilva) [reply src/models/indexer/response-guard.spec.ts]: If we send this data to the backend it'll accept that?

Comment (Puszkarek) [reply src/models/indexer/response-guard.spec.ts]: we don't send, it only has a `getAll`
